packages/core/src/scores/scoreBatchTraces/utils.ts

```diff
@@ -0,0 +1,324 @@
+    // @ts-ignore
+    return input.messages.map(msg => convertToUIMessage(msg, agentSpan.startedAt));
+  }
+  return [];
+}
+
+/**
+ * Extract system messages from LLM span
+ */
+function extractSystemMessages(llmSpan: AISpanRecord): Array<{ role: 'system'; content: string }> {
+  return (llmSpan.input?.messages || [])
+    .filter((msg: any) => msg.role === 'system')
+    .map((msg: any) => ({
+      role: 'system' as const,
+      content: normalizeMessageContent(msg.content),
+    }));
+}
+
+/**
+ * Extract conversation history (remembered messages) from LLM span
+ * Excludes system messages and the current input message
+ */
+function extractRememberedMessages(llmSpan: AISpanRecord, currentInputContent: string): TransformedUIMessage[] {
+  const messages = (llmSpan.input?.messages || [])
+    .filter((msg: any) => msg.role !== 'system')
+    .filter((msg: any) => normalizeMessageContent(msg.content) !== currentInputContent);
+
+  return messages.map((msg: any) => convertToUIMessage(msg, llmSpan.startedAt));
+}
+
+/**
+ * Reconstruct tool invocations from tool call spans
+ */
+function reconstructToolInvocations(spanTree: SpanTree, parentSpanId: string) {
+  const toolSpans = getChildrenOfType<AISpanRecord>(spanTree, parentSpanId, AISpanType.TOOL_CALL);
+
+  return toolSpans.map(toolSpan => ({
+    state: 'result' as const,
+    toolName: toolSpan.attributes?.toolId,
+    args: toolSpan.input || {},
+    result: toolSpan.output || {},
+  }));
+}
+
+/**
+ * Create message parts array including tool invocations and text
+ */
+function createMessageParts(toolInvocations: AISpanRecord[], textContent: string) {
+  const parts: { type: 'tool-invocation' | 'text'; toolInvocation?: AISpanRecord; text?: string }[] = [];
+  for (const toolInvocation of toolInvocations) {
+    parts.push({
+      type: 'tool-invocation',
+      toolInvocation,
+    });
+  }
+
+  if (textContent.trim()) {
+    parts.push({
+      type: 'text',
+      text: textContent,
+    });
+  }
+
+  return parts;
+}
+
+/**
+ * Validate trace structure and throw descriptive errors
+ */
+export function validateTrace(trace: AITraceRecord): void {
+  if (!trace) {
+    throw new Error('Trace is null or undefined');
+  }
+
+  if (!trace.spans || !Array.isArray(trace.spans)) {
+    throw new Error('Trace must have a spans array');
+  }
+
+  if (trace.spans.length === 0) {
+    throw new Error('Trace has no spans');
+  }
+
+  // Check for circular references in parent-child relationships
+  const spanIds = new Set(trace.spans.map(span => span.spanId));
+  for (const span of trace.spans) {
+    if (span.parentSpanId && !spanIds.has(span.parentSpanId)) {
+      throw new Error(`Span ${span.spanId} references non-existent parent ${span.parentSpanId}`);
+    }
+  }
+}
+
+/**
+ * Find the most recent LLM span that contains conversation history
+ */
+function findPrimaryLLMSpan(spanTree: SpanTree, rootAgentSpan: AISpanRecord): AISpanRecord {
```

Comment:
there should only be a single `LLM_GENERATION` span per agentSpan... and it should be a direct child of the `rootAgentSpan`.

if there are others in the trace, those are associated with nested agents.